trunk/SUAVE/Methods/Aerodynamics/Airfoil_Panel_Method/airfoil_analysis.py

```diff
@@ -0,0 +1,573 @@
+    mask_count      = np.ma.count(X_BL_TOP,axis = 0)
+    prev_index      = first_idx-1
+    first_panel     = list(prev_index.flatten())
+    last_panel      = list((first_idx-1 + mask_count).flatten())
+    last_paneldve   = list((first_idx-2 + mask_count).flatten())
+    aoas            = list(np.repeat(np.arange(nalpha),nRe))
+    res             = list(np.tile(np.arange(nRe),nalpha) )
+    X_BL_TOP.mask[first_panel,aoas,res] = False
+    
+    # flow velocity and pressure of on botton surface 
+    VE_BL_TOP    = VT_BL
+    CP_BL_TOP    = 1 - VE_BL_TOP**2     
+    
+    CP_BL_VALS   = np.ma.concatenate([np.flip(CP_BL_TOP,axis = 0),CP_BL_BOT], axis = 0 )  
+    CP_BL_VALS_1 = CP_BL_VALS.flatten('F')  
+    CP_BL_VALS_2 = CP_BL_VALS_1.data[~CP_BL_VALS_1.mask] 
+    CP_BL        = CP_BL_VALS_2.reshape((npanel,nalpha,nRe),order = 'F')     
+    
+    AERO_RES_BL        = aero_coeff(X,Y,-CP_BL,alpha,npanel) 
+    airfoil_properties = Data(
+        AoA        = alpha,
+        Re         = Re_L,
+        Cl         = AERO_RES_BL.Cl,
+        Cd         = AERO_RES_BL.Cd,
+        Cm         = AERO_RES_BL.Cm,  
+        normals    = normals,
+        x          = X,
+        y          = Y,        
+        x_bl       = X_BL,
+        y_bl       = Y_BL,
+        Cp         = CP_BL,         
+        Ue_Vinf    = VE,         
+        dVe        = DVE,   
+        theta      = THETA,      
+        delta_star = DELTA_STAR, 
+        delta      = DELTA,
+        Re_theta   = RE_THETA,
+        Re_x       = RE_X,
+        H          = H,               
+        Cf         = CF,          
+        )  
+    
+    if batch_analysis:
+        pass
+    else:
+        airfoil_properties = extract_values(airfoil_properties) 
+        
+    return  airfoil_properties 
+
+
+def interpolate_values(X_BOT,X_TOP,X_BOT_SURF,X_TOP_SURF,FUNC_BOT_SURF,FUNC_TOP_SURF,npanel,nalpha,nRe,batch_analysis): 
+    '''Interpolation of airfoil properties''' # numpy.ma.apply_along_axis¶
+    FUNC = np.zeros((npanel,nalpha,nRe))
+    
+    if batch_analysis:
+        N_ALPHA = nalpha
+    else:
+        N_ALPHA = 1 
+    
+    for a_i in range(N_ALPHA):
+        for Re_i in range(nRe):    
+            if not batch_analysis:
+                a_i = Re_i  
+            top_func          = interp1d(np.flip(X_TOP_SURF, axis=0)[:,a_i,Re_i],np.flip(FUNC_TOP_SURF, axis=0)[:,a_i,Re_i], fill_value='extrapolate')
+            bot_func          = interp1d(X_BOT_SURF[:,a_i,Re_i],FUNC_BOT_SURF[:,a_i,Re_i], fill_value='extrapolate')  
+            x_top             = np.flip(X_TOP, axis=0)[:,a_i,Re_i]
+            x_top             = x_top[x_top.mask == False].data    
+            x_bot             = X_BOT[:,a_i,Re_i]
+            x_bot             = x_bot[x_bot.mask == False].data                  
+            FUNC[:,a_i,Re_i]  = np.concatenate([top_func(x_top),bot_func(x_bot)])
+    return FUNC
+
+
+def extract_values(AP):    
+    AP.Cl         = np.diagonal(AP.Cl,axis1 = 0, axis2 = 1)
```

Comment:
I wrote the code to be as fast as possible for batch analysis. As a result, the data structure is a 2d matrix rather than a 1d array. The results from an analysis that is not in batch mode i.e., if you want to run a series of different airfoils and different conditions lie on the diagonal. This numpy function simply retrieves the diagonal